### PR TITLE
Be consistent with downloads

### PR DESCRIPTION
**Pull request by baltitenger**
Use wget everywhere, store all temporary files in /tmp/cache.

### CODE REVIEW

**Comment by baltitenger** [comment]
Note: also updated vscode cpptools to v1.16.3.

**Comment by baltitenger** [comment]
I just wanted it to be obvious that it's not running plain wget. I guess it could be an alias with a different name as well.

File: cleanup.sh [hunk 1]
```diff
@@ -4,6 +4,7 @@ set -e
 
 apt clean
 rm -rf /var/lib/snapd/cache/*
+rm -rf /tmp/cache
```
Comment (radl97) [inline cleanup.sh:7]: There are some lines removing specific directories from /tmp, then all of tmp, then some other parts... This is pretty useless, as `/tmp` is stored in RAM (for a very long time now on major distros).

So this line of code is useless twice: it gets deleted 10 lines below, and it is removed already on reboot. (Although, I think that the thing 10 lines below would be happy to break some non-trivial GUI apps)

```bash

# Clear out temporary files
rm -rf /tmp/*
rm -rf /var/tmp/*
```

Maybe revamping this kinds of code could be useful.

File: setup.sh [hunk 2]
```diff
@@ -116,16 +119,15 @@ mkdir -p /opt/ioi/store/screenshots
 mkdir -p /opt/ioi/store/submissions
 mkdir -p /opt/ioi/config/ssh
 
-# Latest as of 2023-05-20
-aria2c -x 4 -d /tmp -o cpptools-linux.vsix "https://github.com/microsoft/vscode-cpptools/releases/download/v1.15.4/cpptools-linux.vsix"
-wget -O /tmp/vscodevim.vsix "https://github.com/VSCodeVim/Vim/releases/download/v1.25.2/vim-1.25.2.vsix"
-rm -rf /tmp/vscode
```
Comment (radl97) [inline setup.sh:122]: This is useful for testing. It was not probably an accident. This change is unrelated to the PR.

File: setup.sh [hunk 3]
```diff
@@ -116,16 +119,15 @@ mkdir -p /opt/ioi/store/screenshots
...
-mkdir /tmp/vscode-extensions
-code --install-extension /tmp/cpptools-linux.vsix --extensions-dir /tmp/vscode-extensions --user-data-dir /tmp/vscode
+# Latest as of 2023-07-23
+$wget -O $cache/cpptools-1.16.3.vsix https://github.com/microsoft/vscode-cpptools/releases/download/v1.16.3/cpptools-linux.vsix
+mkdir /tmp/vscode /tmp/vscode-extensions
+code --install-extension $cache/cpptools-1.16.3.vsix --extensions-dir /tmp/vscode-extensions --user-data-dir /tmp/vscode
 tar jcf /opt/ioi/misc/vscode-extensions.tar.bz2 -C /tmp/vscode-extensions .
-cp /tmp/vscodevim.vsix /opt/ioi/misc
-rm -rf /tmp/vscode-extensions
+rm -rf /tmp/vscode /tmp/vscode-extensions
```
Comment (radl97) [inline setup.sh:127]: I think we do not want to delete directories in the setup script. It seemed that the idea all along was that if anything goes bad in the setup, we need to investigate it and understand where it went wrong. Hence there are very few deletions whatsoever.

Adding `/tmp/vscode` is an unrelated change in the PR.

File: setup.sh [hunk 4]
```diff
@@ -202,10 +204,9 @@ apt -y install python3-doc
 
 # CPP Reference
 
-wget -O /tmp/html_book.zip https://github.com/PeterFeicht/cppreference-doc/releases/download/v20220730/html-book-20220730.zip
+$wget https://github.com/PeterFeicht/cppreference-doc/releases/download/v20220730/html-book-20220730.zip
 mkdir -p /usr/share/doc/cppreference
-unzip -o /tmp/html_book.zip -d /usr/share/doc/cppreference
-rm -f /tmp/html_book.zip
+unzip -o $cache/html-book-20220730.zip -d /usr/share/doc/cppreference
```
Comment (radl97) [inline setup.sh:209]: There is a duplication of version here. It was probably intended to download to a fix location above:

```bash
wget -O /tmp/html_book.zip https://github.com/PeterFeicht/cppreference-doc/releases/download/v20220730/html-book-20220730.zip
```

File: cleanup.sh (hunk 1 above)
Comment (radl97) [reply cleanup.sh:7]: I would remove this line.

Comment (baltitenger) [reply cleanup.sh:7]: I might recall wrong but last time I checked it seemd like /tmp wasn't tmpfs.

File: setup.sh (hunk 4 above)
Comment (baltitenger) [reply setup.sh:209]: The wget command I use will not download the file if it already exists. That's why I always include a version number in the local filename as well. (otherwise if the url changed but the cache file stays then it won't download the new version)

Perhaps there's a better way to do this but this is all I managed to come up with.

File: cleanup.sh (hunk 1 above)
Comment (baltitenger) [reply cleanup.sh:7]: Actually, if /tmp were a tmpfs I think we shouldn't use it for storing such large files.

File: setup.sh (hunk 4 above)
Comment (radl97) [reply setup.sh:209]: I think reentrancy (allowing `setup.sh` to run multiple times) would be a nice-to-have feature, simply for debugging.

Comment (radl97) [reply setup.sh:209]: Meanwhile, I understand the idea behind this change.

**Comment by baltitenger** [comment]
My main use case for all this is that I have a persistent disk mounted during setup.sh so even if I rebuild the image I basically don't have to download anything.

(I have a separate script for setting up the mounts, but I didn't add it to the repo as it might be regarded as a bit of a hack. I mostly used it while I was on holiday and had a rather slow connection.)

File: cleanup.sh (hunk 1 above)
Comment (radl97) [reply cleanup.sh:7]: Yeah, it's not tmpfs, sorry. I have many problems right now accepting this :D How did I not notice in the previous N years?

However, the main point still stands: `/tmp/cache` is already deleted somewhere else.

Comment (baltitenger) [reply cleanup.sh:7]: Fair enough.

Comment (radl97) [reply cleanup.sh:7]: No, because it's generally run at bootup, not at shutdown (and problem is that we are creating an image without starting up the system)

File: setup.sh (hunk 4 above)
Comment (radl97) [reply setup.sh:209]: Actually... Yes, I've been thinking that it would be an upgrade. Either ansible or cloud-init. (latter already set up and supported, but not required)

**Comment by radl97** [comment]
It's a great work you did there.

However, based on the changes, I opened #61 to for minor changes regarding the patches (revert some, document some, and split to more commits). I hope it's not a problem.
